### PR TITLE
[ISSUE-492] Manipulate Drive Usage via annotations

### PR DESCRIPTION
**Pull request by safronovD**
### Issue #492 

Annotations for drive CR to change Usage:
1. action=add
FAILED -> IN_USE (repeat drive replacement procedure)
RELEASED -> IN_USE (in RELEASED state user can continue DR with `replacement=ready` or rollback, if it was started by mistake)
2. action=remove
FAILED -> REMOVED (user can finish DR for FAILED disk)

Additional:
1. Nullify AC for disk, which Usage isn't IN_USE

## PR checklist
- [x] Add link to the issue
- [x] Choose PR label
- [ ] New unit tests added
- [ ] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

### Remove FAILED disk
```
csi-rke-sles15-104:~ # kubectl get drive -o wide 6ef7e16a-bc41-4746-b3f9-13402ba209d8 
NAME                                   SERIAL NUMBER                      HEALTH   TYPE   NODE                                   SIZE           SLOT   USAGE    STATUS
6ef7e16a-bc41-4746-b3f9-13402ba209d8   6000C29B1AFCEF54D337781E7B2621CC   GOOD     HDD    22c58825-0a8a-4700-84ec-b83ba8f713c8   241000000000   2      IN_USE   ONLINE
csi-rke-sles15-104:~ # kubectl annotate drive 6ef7e16a-bc41-4746-b3f9-13402ba209d8 health=bad
drive.csi-baremetal.dell.com/6ef7e16a-bc41-4746-b3f9-13402ba209d8 annotated
csi-rke-sles15-104:~ # kubectl get drive -o wide 6ef7e16a-bc41-4746-b3f9-13402ba209d8 
NAME                                   SERIAL NUMBER                      HEALTH   TYPE   NODE                                   SIZE           SLOT   USAGE      STATUS
6ef7e16a-bc41-4746-b3f9-13402ba209d8   6000C29B1AFCEF54D337781E7B2621CC   BAD      HDD    22c58825-0a8a-4700-84ec-b83ba8f713c8   241000000000   2      RELEASED   ONLINE
csi-rke-sles15-104:~ # kubectl annotate drive 6ef7e16a-bc41-4746-b3f9-13402ba209d8 replacement=ready
drive.csi-baremetal.dell.com/6ef7e16a-bc41-4746-b3f9-13402ba209d8 annotated
csi-rke-sles15-104:~ # kubectl get drive -o wide 6ef7e16a-bc41-4746-b3f9-13402ba209d8 
NAME                                   SERIAL NUMBER                      HEALTH   TYPE   NODE                                   SIZE           SLOT   USAGE    STATUS
6ef7e16a-bc41-4746-b3f9-13402ba209d8   6000C29B1AFCEF54D337781E7B2621CC   BAD      HDD    22c58825-0a8a-4700-84ec-b83ba8f713c8   241000000000   2      FAILED   ONLINE
csi-rke-sles15-104:~ # kubectl annotate drive 6ef7e16a-bc41-4746-b3f9-13402ba209d8 action=remove
drive.csi-baremetal.dell.com/6ef7e16a-bc41-4746-b3f9-13402ba209d8 annotated
csi-rke-sles15-104:~ # kubectl get drive -o wide 6ef7e16a-bc41-4746-b3f9-13402ba209d8 
NAME                                   SERIAL NUMBER                      HEALTH   TYPE   NODE                                   SIZE           SLOT   USAGE     STATUS
6ef7e16a-bc41-4746-b3f9-13402ba209d8   6000C29B1AFCEF54D337781E7B2621CC   BAD      HDD    22c58825-0a8a-4700-84ec-b83ba8f713c8   241000000000   2      REMOVED   ONLINE
csi-rke-sles15-104:~ # lsblk -o NAME,HCTL | grep sdc
sdc             0:0:2:0
csi-rke-sles15-104:~ # echo 1 | sudo tee /sys/class/scsi_device/0:0:2:0/device/delete
1
csi-rke-sles15-104:~ # lsblk -o NAME,HCTL | grep sdc
csi-rke-sles15-104:~ # kubectl get drive -o wide 6ef7e16a-bc41-4746-b3f9-13402ba209d8 
Error from server (NotFound): drives.csi-baremetal.dell.com "6ef7e16a-bc41-4746-b3f9-13402ba209d8" not found
csi-rke-sles15-104:~ # kubectl get events --sort-by='.lastTimestamp' | grep 6ef7e16a-bc41-4746-b3f9-13402ba209d8
10m         Normal    DriveDiscovered            drive/6ef7e16a-bc41-4746-b3f9-13402ba209d8            New drive discovered SN: 6000C29B1AFCEF54D337781E7B2621CC, Node: 22c58825-0a8a-4700-84ec-b83ba8f713c8. Drive Details: SN='6000C29B1AFCEF54D337781E7B2621CC', Node='22c58825-0a8a-4700-84ec-b83ba8f713c8', Type='HDD', Model='VMware Virtual disk', Size='241000000000', Firmware=''
10m         Normal    DriveHealthGood            drive/6ef7e16a-bc41-4746-b3f9-13402ba209d8            Drive health is: GOOD, previous state: UNKNOWN. Drive Details: SN='6000C29B1AFCEF54D337781E7B2621CC', Node='22c58825-0a8a-4700-84ec-b83ba8f713c8', Type='HDD', Model='VMware Virtual disk', Size='241000000000', Firmware=''
7m57s       Warning   DriveHealthOverridden      drive/6ef7e16a-bc41-4746-b3f9-13402ba209d8            Drive health is overridden with: BAD, real state: GOOD. Drive Details: SN='6000C29B1AFCEF54D337781E7B2621CC', Node='22c58825-0a8a-4700-84ec-b83ba8f713c8', Type='HDD', Model='VMware Virtual disk', Size='241000000000', Firmware=''
7m57s       Error     DriveHealthFailure         drive/6ef7e16a-bc41-4746-b3f9-13402ba209d8            Drive health is: BAD, previous state: GOOD. Drive Details: SN='6000C29B1AFCEF54D337781E7B2621CC', Node='22c58825-0a8a-4700-84ec-b83ba8f713c8', Type='HDD', Model='VMware Virtual disk', Size='241000000000', Firmware=''
7m56s       Warning   DriveReadyForRemoval       drive/6ef7e16a-bc41-4746-b3f9-13402ba209d8            Drive is ready for removal, Drive Details: SN='6000C29B1AFCEF54D337781E7B2621CC', Node='22c58825-0a8a-4700-84ec-b83ba8f713c8', Type='HDD', Model='VMware Virtual disk', Size='241000000000', Firmware=''
2m26s       Error     DriveRemovalFailed         drive/6ef7e16a-bc41-4746-b3f9-13402ba209d8            Failed to locale LED, Drive Details: SN='6000C29B1AFCEF54D337781E7B2621CC', Node='22c58825-0a8a-4700-84ec-b83ba8f713c8', Type='HDD', Model='VMware Virtual disk', Size='241000000000', Firmware=''
2m2s        Warning   DriveRemovedByForce        drive/6ef7e16a-bc41-4746-b3f9-13402ba209d8            Drive was removed via annotation, Drive Details: SN='6000C29B1AFCEF54D337781E7B2621CC', Node='22c58825-0a8a-4700-84ec-b83ba8f713c8', Type='HDD', Model='VMware Virtual disk', Size='241000000000', Firmware=''
24s         Warning   DriveHealthOverridden      drive/6ef7e16a-bc41-4746-b3f9-13402ba209d8            Drive health is overridden with: UNKNOWN, real state: BAD. Drive Details: SN='6000C29B1AFCEF54D337781E7B2621CC', Node='22c58825-0a8a-4700-84ec-b83ba8f713c8', Type='HDD', Model='VMware Virtual disk', Size='241000000000', Firmware=''
24s         Normal    DriveSuccessfullyRemoved   drive/6ef7e16a-bc41-4746-b3f9-13402ba209d8            Drive successfully removed. Drive status is: OFFLINE, previous status: ONLINE. Drive Details: SN='6000C29B1AFCEF54D337781E7B2621CC', Node='22c58825-0a8a-4700-84ec-b83ba8f713c8', Type='HDD', Model='VMware Virtual disk', Size='241000000000', Firmware=''
```

### Return disk after RELEASED
```
user@user-vm ~/g/s/g/dell> k annotate --overwrite drive 16f8c148-5ade-4a55-be3c-40515ee614e4 health=bad
drive.csi-baremetal.dell.com/16f8c148-5ade-4a55-be3c-40515ee614e4 annotated
user@user-vm ~/g/s/g/dell> kgo drive 16f8c148-5ade-4a55-be3c-40515ee614e4
NAME                                   SERIAL NUMBER        HEALTH   TYPE   NODE                                   SIZE        SLOT   USAGE      STATUS
16f8c148-5ade-4a55-be3c-40515ee614e4   LOOPBACK2298887491   BAD      HDD    d59f49a6-05e8-4752-bbdb-af9e925a06f6   105906176          RELEASED   ONLINE
user@user-vm ~/g/s/g/dell> kg ac 874cd02a-c8e9-4ee4-8637-846c55cca92
NAME                                   LOCATION                               NODE                                   STORAGE CLASS   SIZE
874cd02a-c8e9-4ee4-8637-846c55cca922   16f8c148-5ade-4a55-be3c-40515ee614e4   d59f49a6-05e8-4752-bbdb-af9e925a06f6   HDD             
user@user-vm ~/g/s/g/dell> k annotate --overwrite drive 16f8c148-5ade-4a55-be3c-40515ee614e4 health=-
drive.csi-baremetal.dell.com/16f8c148-5ade-4a55-be3c-40515ee614e4 annotated
user@user-vm ~/g/s/g/dell> kgo drive 16f8c148-5ade-4a55-be3c-40515ee614e4
NAME                                   SERIAL NUMBER        HEALTH   TYPE   NODE                                   SIZE        SLOT   USAGE      STATUS
16f8c148-5ade-4a55-be3c-40515ee614e4   LOOPBACK2298887491   GOOD     HDD    d59f49a6-05e8-4752-bbdb-af9e925a06f6   105906176          RELEASED   ONLINE
user@user-vm ~/g/s/g/dell> k annotate --overwrite drive 16f8c148-5ade-4a55-be3c-40515ee614e4 action=add
drive.csi-baremetal.dell.com/16f8c148-5ade-4a55-be3c-40515ee614e4 annotated
user@user-vm ~/g/s/g/dell> kgo drive 16f8c148-5ade-4a55-be3c-40515ee614e4
NAME                                   SERIAL NUMBER        HEALTH   TYPE   NODE                                   SIZE        SLOT   USAGE    STATUS
16f8c148-5ade-4a55-be3c-40515ee614e4   LOOPBACK2298887491   GOOD     HDD    d59f49a6-05e8-4752-bbdb-af9e925a06f6   105906176          IN_USE   ONLINE
user@user-vm ~/g/s/g/dell> kg ac 874cd02a-c8e9-4ee4-8637-846c55cca922
NAME                                   LOCATION                               NODE                                   STORAGE CLASS   SIZE
874cd02a-c8e9-4ee4-8637-846c55cca922   16f8c148-5ade-4a55-be3c-40515ee614e4   d59f49a6-05e8-4752-bbdb-af9e925a06f6   HDD             105906176
```

### CODE REVIEW

**Review by yurkov-anton — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -103,12 +103,13 @@ generate-deepcopy:
 	controller-gen object paths=api/v1/nodecrd/node_types.go paths=api/v1/nodecrd/groupversion_info.go  output:dir=api/v1/nodecrd
 
 generate-crds:
-    # Generate CRDs based on Volume and AvailableCapacity type and group info
```
Comment (mishoyama) [inline Makefile:106]: What problem are you solving here?

File: pkg/crcontrollers/drive/drivecontroller.go [hunk 2]
```diff
@@ -40,9 +40,10 @@ const (
 )
 
 const (
-	// Annotations for driveCR to perform restart process
-	driveRestartReplacementAnnotationKey   = "drive"
```
Comment (mishoyama) [inline pkg/crcontrollers/drive/drivecontroller.go:44]: Don't we need to support this key as deprecated and remove later?

File: pkg/crcontrollers/drive/drivecontroller.go [hunk 3]
```diff
@@ -284,14 +279,46 @@ func (c *Controller) checkAllVolsRemoved(volumes []*volumecrd.Volume) bool {
 	return true
 }
 
-// restartReplacement restores drive.Usage to IN_USE if CR is annotated
+// placeStatusInUse places drive.Usage to IN_USE if CR is annotated
 // deletes the annotation to avoid event repeating
-func (c *Controller) restartReplacement(drive *drivecrd.Drive) bool {
-	if value, ok := drive.GetAnnotations()[driveRestartReplacementAnnotationKey]; ok && value == driveRestartReplacementAnnotationValue {
+func (c *Controller) placeStatusInUse(drive *drivecrd.Drive) bool {
```
Comment (mishoyama) [inline pkg/crcontrollers/drive/drivecontroller.go:284]: Please rename to `checkAndPlace...`

File: pkg/crcontrollers/drive/drivecontroller.go [hunk 4]
```diff
@@ -284,14 +279,46 @@ func (c *Controller) checkAllVolsRemoved(volumes []*volumecrd.Volume) bool {
...
+		delete(drive.Annotations, driveActionAnnotationKey)
 		return true
 	}
 
 	return false
 }
+
+// placeStatusRemoved places drive.Usage to REMOVED if CR is annotated
+// deletes the annotation to avoid event repeating
+func (c *Controller) placeStatusRemoved(drive *drivecrd.Drive) bool {
```
Comment (mishoyama) [inline pkg/crcontrollers/drive/drivecontroller.go:296]: Same here

File: pkg/crcontrollers/drive/drivecontroller.go [hunk 5]
```diff
@@ -229,7 +224,7 @@ func (c *Controller) handleDriveUpdate(ctx context.Context, log *logrus.Entry, d
 			return remove, nil
 		}
 	case apiV1.DriveUsageFailed:
```
Comment (mishoyama) [inline pkg/crcontrollers/drive/drivecontroller.go:226]: From failed status we need to be able to move to the REMOVED state.

**Review by mishoyama — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (safronovD) [reply Makefile:106]: It makes easier to copy crds into csi-baremetal-operator chart and give the opportunity to check in pr-validation that we have actual files in our repo (yamls in commit equals to `make generate-crds` output). Also we don't support these charts, it was confusing for me

File: pkg/crcontrollers/drive/drivecontroller.go (hunk 2 above)
Comment (safronovD) [reply pkg/crcontrollers/drive/drivecontroller.go:44]: add annotation key&value as deprecated

File: pkg/crcontrollers/drive/drivecontroller.go (hunk 3 above)
Comment (safronovD) [reply pkg/crcontrollers/drive/drivecontroller.go]: renamed

File: pkg/crcontrollers/drive/drivecontroller.go (hunk 4 above)
Comment (safronovD) [reply pkg/crcontrollers/drive/drivecontroller.go]: renamed

File: pkg/crcontrollers/drive/drivecontroller.go (hunk 5 above)
Comment (safronovD) [reply pkg/crcontrollers/drive/drivecontroller.go:229]: We check and change state to REMOVED before main switch. It works for any state

File: api/generated/crds/csi-baremetal.dell.com_availablecapacities.yaml [hunk 6]
```diff
@@ -0,0 +1,78 @@
+
```
Comment (mishoyama) [inline api/generated/crds/csi-baremetal.dell.com_availablecapacities.yaml:1]: What is the reason to keep these crds in 2 repos at the same time?

File: Makefile (hunk 1 above)
Comment (mishoyama) [reply Makefile:106]: I recommend you to not mix 2 different tasks in single pr. Please open a separate issue for charts modifications

File: pkg/crcontrollers/drive/drivecontroller.go (hunk 5 above)
Comment (mishoyama) [reply pkg/crcontrollers/drive/drivecontroller.go:229]: What is the use case?

File: Makefile (hunk 1 above)
Comment (safronovD) [reply Makefile]: Changes were removed from this pr, I will open another

File: pkg/crcontrollers/drive/drivecontroller.go (hunk 5 above)
Comment (safronovD) [reply pkg/crcontrollers/drive/drivecontroller.go:229]: Force disk removing, if disk is empty or it's already OFFLINE. Is it better leave only FAILED -> REMOVED?

File: api/generated/crds/csi-baremetal.dell.com_availablecapacities.yaml (hunk 6 above)
Comment (safronovD) [reply api/generated/crds/csi-baremetal.dell.com_availablecapacities.yaml]: Discussed on the meeting. I will create issue for it

**Pull request merged**